File: wrapper/toolchain_wrapper.py
import sys
import os
from sys import exit

if __name__ == '__main__':
    os.environ['PATH'] = os.path.dirname(__file__) + os.path.pathsep + os.environ['PATH']
    # Set path of CA certs
    os.environ['SSL_CERT_FILE'] = os.path.join(os.path.dirname(__file__), 'cacert.pem')
    if '__ARG0' in os.environ:
        sys.argv[0] = os.environ['__ARG0']

    basename = os.path.splitext(os.path.basename(sys.argv[0]))[0]
    target = basename.rsplit('-', maxsplit=1)[0]
    exec_name = basename.rsplit('-', maxsplit=1)[-1]
    if target == basename:
        sys.stderr.write('toolchain-wrapper: cannot run tools without target triple prefix\n')
        exit(1)
    if exec_name in ('clang', 'clang++', 'cc', 'c++', 'gcc', 'g++', 'as', 'cl'):
        import clang_wrapper
        clang_wrapper.main(target, exec_name)
    elif 'apple-' in target:
        import cctools_wrapper
        cctools_wrapper.main(target, exec_name)
    elif exec_name in ('ld', 'link'):
        import lld_wrapper
        lld_wrapper.main(target, exec_name)
    else:
        sys.stderr.write('toolchain-wrapper: cannot find correspond wrapper for %s\n' % basename)
        exit(1)
    

File: wrapper/toolchain_wrapper.go
package main

import (
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type StringStringPair struct {
	first  string
	second string
}

func inArray(arr []string, target string) bool {
	for _, val := range arr {
		if val == target {
			return true
		}
	}
	return false
}

func runCommand(execPath string, argv []string, inputBytes []byte) {
	cmd := exec.Command(execPath, argv...)
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	var err error = nil
	var stdin io.WriteCloser = nil
	if len(inputBytes) == 0 {
		cmd.Stdin = os.Stdin
	} else {
		stdin, err = cmd.StdinPipe()
		if err != nil {
			panic(err)
		}
	}
	err = cmd.Start()
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	if stdin != nil {
		stdin.Write(inputBytes)
		stdin.Close()
	}
	cmd.Wait()
	os.Exit(cmd.ProcessState.ExitCode())
}

func main() {
	execDir, err := os.Executable()
	if err != nil {
		panic(err)
	}
	execDir = filepath.Dir(execDir)
	os.Setenv("PATH", execDir+string(os.PathListSeparator)+os.Getenv("PATH"))
	basename := filepath.Base(os.Args[0])
	basename = strings.TrimSuffix(basename, filepath.Ext(basename))
	_tempSplit := strings.Split(basename, "-")
	execName := _tempSplit[len(_tempSplit)-1]
	target := strings.TrimSuffix(basename, "-"+execName)
	if target == basename {
		fmt.Fprintln(os.Stderr, "toolchain-wrapper: cannot run tools without target triple prefix")
		os.Exit(1)
	}
	if inArray([]string{"clang", "clang++", "cc", "c++", "gcc", "g++", "as", "cpp", "cl"}, execName) {
		clangWrapperMain(execDir, target, execName, os.Args)
	} else if strings.Contains(target, "apple-") {
		ccToolsWrapperMain(execDir, target, execName, os.Args)
	} else if inArray([]string{"ld", "link"}, execName) {
		lldWrapperMain(execDir, target, execName, os.Args)
	} else {
		fmt.Fprintln(os.Stderr, "toolchain-wrapper: cannot find correspond wrapper for "+basename)
		os.Exit(1)
	}
}
